### ISSUE TITLE
Unit Testing

### ISSUE THREAD

**Issue by retailcoder** [WaynePhillipsEA/twinbasic#54]
**Is your feature request related to a problem? Please describe.**

Testable code is nice, *tested* code is better. The VSC plug-in needs a way to discover and run twinBASIC unit tests.

**Describe the solution you'd like**

- A command-line tool that can load the VBA source files in a folder, compile the project, discover the tests, run them, and report back with the results.

- twinBASIC source code being built and tested on `{insert cloud CI service}` when a pull request is merged into a twinBASIC repository on GitHub, for example.

**Describe alternatives you've considered**

VB6 and VBA code (and thus back-compatible twinBASIC code) can be unit-tested in the VBE with Rubberduck, but the add-in isn't the compiler and only Rubberduck can command a test run, and well it's complicated. It's not really an alternative, because twinBASIC needs its own test runner for any code that *isn't* back-compatible.

Could Rubberduck then somehow start leveraging a twinBASIC API to discover, build & run the VBA/VB6 tests, and get the results? Oh my, the *realm of possibilities* if these two streams ever cross...

**Comment by WaynePhillipsEA** [comment]
My initial plan on this is to first implement the official testing API from VSCode.  Unfortunately, it's not yet finished:
https://code.visualstudio.com/api/extension-guides/testing
So we're currently blocked by 
https://github.com/microsoft/vscode/issues/107467

As soon as that is incorporated, this can be added quickly.  Need to think a bit more about how the Ducky could leverage it.

**Comment by WaynePhillipsEA** [comment]
We should also discuss the discoverability of unit tests here.  I assume supporting the Rubberduck annotations would be the preferred option?

**Comment by retailcoder** [comment]
Rubberduck is requiring a `@TestModule` annotation in any standard module to start iterating the members of a module, then procedures annotated with `@TestMethod` are discovered as unit tests.

There were name/convention-based discovery mechanisms in earlier versions (e.g. any public sub in a test module with a name that starts with "Test"), but it was dropped in favor of explicitness.

**Comment by bclothier** [comment]
I should point out that goes back to the discussion on attributes in #25 . I would encourage that the question of attributes and its syntax be settled. It can then "auto-translate" Rubberduck's annotations if needed to. 

In .NET, attributes are themselves code, while Rubberduck's annotations in VBA are just a special kind of comment and thus not code in the strict sense. I think there's an advantage in having attributes as code. VB6/VBA's `Attribute` keyword could be overloaded but I admit I'm hesitant to write `Attribute TestMethod` because it's verbose and may also break back-compat with V6/VBA. Alternatively, twinBASIC could choose to treat `'@` prefix as attribute and treat the rest as _code_, which would then backport the RD's annotations without affecting the back-compat.

**Comment by WaynePhillipsEA** [comment]
It looks like the vscode Testing API is now being finalized, so we should be able to get onto this soon!

**Comment by WaynePhillipsEA** [comment]
Any thoughts on how we want the attributes to look for unit testing?  I'm hoping the VS Code feature will be finalized for the June update, so we should start to plan for this.

**Comment by bclothier** [comment]
I'm generally to copy from others that has worked well. In Rubberduck's case, we use NUnit as this works very well for its needs. We used to use MSTest but it was a bit clunky. 

Therefore, borrowing from NUnit would probably be a good start. They have a [page listing all attributes used by their engine](https://docs.nunit.org/articles/nunit/writing-tests/attributes.html).

**Comment by WaynePhillipsEA** [comment]
Unfortunately the finalization of the VS code testing API was pushed back another month, so should now be ready for the July release.

**Comment by WaynePhillipsEA** [comment]
Rather than wait another month, I think I'll get the insiders build of VS code installed next week, and start putting things together for tB support of unit tests, in the hope that they don't change the testing API too much before it is finalized.

**Comment by WaynePhillipsEA** [comment]
Unfortunately the testing API finalization has been put back another month by the VS Code team.

**Comment by WaynePhillipsEA** [comment]
This issue is no longer blocked 😊   With yesterdays update, VS Code now offers an integrated testing API.   Time to get to work...

**Comment by WaynePhillipsEA** [comment]
Throwing this one out to you guys;  for the Assert members such as AreEqual, are we wanting typical VB style equality checks (for example, allowing for default member evaluation and coercions), or are we wanting more stricter checks?   

My personal view is to go for much stricter checks, because we can already do things like 'Debug.Assert A = B' if we want normal VB-style equality checks.

**Comment by mansellan** [comment]
Could this be either an overload or an optional boolean on the Assert methods? I agree that most of the time one would want to use the stricter version, but I can envisage tests which need to match legacy behaviour.

**Comment by WaynePhillipsEA** [comment]
Thanks @mansellan.   I'm introducing an Enum for it called tbAssertStrictness with options of tbAssertStrict, tbAssertExact, tbAssertPermissive.

tbAssertStrict -> no default member evaluations  (DEFAULT)
tbAssertExact -> as tbAssertStrict PLUS no coercions at all (e.g. 5% not equal to 5&)
tbAssertPermissive -> as per normal VBx comparisons

**Comment by WaynePhillipsEA** [comment]
Here's a sneak preview of unit testing in action (rubberduck style attributes for now)

![tb_unitTestsComingSoon](https://user-images.githubusercontent.com/25738109/130360090-da32dac3-9fed-4f47-a62d-05b355c0de84.gif)

Apologies, as the mouse pointer didn't get captured.

**Comment by WaynePhillipsEA** [comment]
**Feedback wanted please!!**

Unit testing is now switched on in v0.10.4880, but considered experimental for the moment. 

#### Limitations
- only supports module based unit tests at the moment
- there are no setup/teardown methods yet

#### Usage
- add [TestFixture] attribute before a module definition
- add [TestCase] attribute before each test procedure
- use the Assert.Exact / Assert.Strict / Assert.Permissive assertions in your tests
- assertions supported:  AreEqual / AreNotEqual / IsTrue / IsFalse / IsNothing / IsNotNothing / IsNull / IsNotNull
-    plus Succeed / Fail / Inconclusive methods

<img width="724" alt="tb_UnitTests1" src="https://user-images.githubusercontent.com/25738109/130476259-362edeb2-8a84-4577-9466-9954bcfac09b.png">

The VS Code Test Explorer should appear in the activity bar:
<img width="132" alt="tb_UnitTests2" src="https://user-images.githubusercontent.com/25738109/130477005-fe64c32f-9ab0-42d3-b92d-7c6da93bfa73.png">

VS Code offers a few different ways to run tests (singular, group, all):
<img width="367" alt="tb_UnitTests3" src="https://user-images.githubusercontent.com/25738109/130477094-bb2237f9-7892-4d7a-afa3-b48f5b5e3555.png">

## Rules
<img width="228" alt="tb_UnitTests4_Exact" src="https://user-images.githubusercontent.com/25738109/130477281-4743ce15-6fdb-4b5d-b9d7-22f70928ad2a.png">

<img width="242" alt="tb_UnitTests4_Strict" src="https://user-images.githubusercontent.com/25738109/130477216-59abd965-11e5-49d5-8588-55920b75e08e.png">

<img width="248" alt="tb_UnitTests4_Permissive" src="https://user-images.githubusercontent.com/25738109/130477297-0ce87cf5-97d5-40b2-bc54-c60de8c14068.png">

**Comment by Kr00l** [comment]
The first Assert line is invalid but the second one is. Looks like a bug ?
![image](https://user-images.githubusercontent.com/33756666/130567768-5000ddd4-1a5f-444c-94ed-879cd9e88a56.png)

**Comment by WaynePhillipsEA** [comment]
@Kr00l that's fixed in v0.10.4909, thanks.

**Comment by Kr00l** [comment]
What is the sense of the Succeed method in Assert ? For me it does nothing..

**Comment by WaynePhillipsEA** [comment]
> What is the sense of the Succeed method in Assert ? For me it does nothing..

I copied the methods from the prior-art of Rubberduck:  https://github.com/rubberduck-vba/Rubberduck/wiki/Unit-Testing#the-assert-class

Indeed, it does nothing.  Perhaps a visual indication of the expected control flow when writing tests.

I guess an alternative use case might be to have tests be considered Inconclusive unless they explicitly reach the Succeed method, but as far as I know Rubberduck doesn't do that.

**Comment by WaynePhillipsEA** [comment]
v0.10.5135 adds support for Assert members of SequenceEquals and NotSequenceEquals.

**Comment by Kr00l** [comment]
One thing crossing my mind. Is it planned or does it make sense to carve out this unit testing from VBA lib into an own tB lib to tidy things up?

**Comment by WaynePhillipsEA** [comment]
@Kr00l, yes I think that's a good idea, and would also help tidy up the implementation.

**Comment by WaynePhillipsEA** [comment]
As of v0.10.5165, the Assert modules are now only available if explicitly referenced in the project Settings:

<img width="402" alt="tb_AssertLibrary" src="https://user-images.githubusercontent.com/25738109/131297184-20c26d5c-e900-41d0-b8ec-2f8601c43325.png">

_This library does not add any dependencies to the compiled project, as assertions are only evaluated whilst debugging._

**Comment by Kr00l** [comment]
Only AreSame/AreNotSame are missing, or?

**Comment by WaynePhillipsEA** [comment]
Yes, for AreSame/AreNotSame I first want to fix #154 so we can just query for the root IUnknown interface of an object and compare the pointers. For non-tB (and non-VB objects), you could of course get two different object pointers to the 'same' object via distinct interfaces (like tB does now, and which would fail the expected AreSame/AreNotSame assertions), but I think this is acceptable as there is simply no 'perfect' solution to the problem of knowing for sure if two object interfaces refer to the same underlying object.

**Comment by wqweto** [comment]
Getting different IUnknown pointers from two different interfaces of an instance would break COM’s identity rule so don’t think anyone does it while getting different IDispatch impl from these same two interfaces is ok and even desirable so that both set of methods (both interfaces) to be accessible from VBScript or other “scripting” languages based on late-bound calls.

**Comment by WaynePhillipsEA** [comment]
> so don’t think anyone does it

Oh I'm sure plenty do it wrong... but doesn't make it right 😉  Good to know that the identity rule is officially documented.

**Comment by wqweto** [comment]
Interesting, do you have a test-case where VBx does not follow identity rule? 

Last time I cheched VB6 multiple interface impl on a class the ObjPtr on both interfaces’ IUnknown casts matched while IDispatch pointers addresses did differ (and only the corresponding subset of methods were available for late-bound calls).

**Comment by WaynePhillipsEA** [comment]
> Interesting, do you have a test-case where VBx does not follow identity rule?

No, I didn't say VBx's implementation 😉

**Comment by wqweto** [comment]
Btw, do you impl IDispatch::Invoke by forwarding to DispInvoke API with a staticly defined ITypeInfo (a glorified nested struct actually) for the particular dual interface?

I’ve seen IDispatch impl that do not go this route that fail on exotic cases of late-bound calls e.g. IPictureDisp::Render and XMLHttpRequest::Send having troubles with VT_BYREF flagged Variants.

**Comment by WaynePhillipsEA** [comment]
> Btw, do you impl IDispatch::Invoke by forwarding to DispInvoke API with a staticly defined ITypeInfo (a glorified nested struct actually) for the particular dual interface?
> 
> I’ve seen IDispatch impl that do not go this route that fail on exotic cases of late-bound calls e.g. IPictureDisp::Render and XMLHttpRequest::Send having troubles with VT_BYREF flagged Variants.

No, it's a custom implementation because we have to support overloaded members (...although that's not yet supported).  This has a side-benefit of being cross-platform ready.  VT_BYREF is supported, though there is of course potential for edge cases until we have full test coverage.

**Comment by wqweto** [comment]
I see. I thought you were reimplementing DispInvoke on non-Windows OSes but relying on tested COM stdlibs on win32/x64.

How do you plan on supporting overloads in TLB files provided that two methods with same name but different number of parameters will probably break most of the COM consumers of such type library.

You might be forced to differentiate on method name too the way .Net COM interop does it with backtick 2, 3, etc. so the custom Invoke impl might not be enough to pull this trick up :))

**Comment by WaynePhillipsEA** [comment]
> How do you plan on supporting overloads in TLB files provided that two methods with same name but different number of parameters will probably break most of the COM consumers of such type library.

We don't allow that.  Only Private (non-exposed) tB classes can use overloads and generics.

**Comment by wqweto** [comment]
Aha, I see. Neat trick!

So you rely on private interfaces not having to participate in final (public) TLB but now missing ITypeInfo on these interfaces back-fires in some compatibility edge-cases Krool is digging up.

**Comment by WaynePhillipsEA** [comment]
@wqweto I'm not overly concerned with regards to the runtime ITypeInfo support, as not much uses it anyway (outside of marshalling).  I also know that the runtime implementations of ITypeInfo provided by VBA for its classes has **_many_** bugs and quirks that I had to overcome in some of the work I did for Rubberduck... and the tB implementation will be far more conformant than that.  Even if we had to drop overloaded members from the exposed ITypeInfo, it really wouldn't be a big issue.

**Comment by WaynePhillipsEA** [comment]
AreSame and AreNotSame are now implemented, as of v0.10.5276, and the intellisense/signature-help now shows a little help text for each Assert member.

**Comment by WaynePhillipsEA** [comment]
@FullValueRider that's fixed now in v0.11.99, thanks

**Comment by WaynePhillipsEA** [comment]
I'm closing this issue now, as unit testing has been in experimental status for a while now with no recent issues being reported.  For any new problems identified, please open a new issue.  Thanks all.
